server/src/main/java/org/opensearch/cluster/action/shard/LocalShardStateAction.java

```diff
@@ -0,0 +1,82 @@
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.cluster.action.shard;
+
+import org.opensearch.cluster.ClusterState;
+import org.opensearch.cluster.routing.IndexRoutingTable;
+import org.opensearch.cluster.routing.IndexShardRoutingTable;
+import org.opensearch.cluster.routing.RerouteService;
+import org.opensearch.cluster.routing.RoutingTable;
+import org.opensearch.cluster.routing.ShardRouting;
+import org.opensearch.cluster.routing.allocation.AllocationService;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.inject.Inject;
+import org.opensearch.core.action.ActionListener;
+import org.opensearch.threadpool.ThreadPool;
+import org.opensearch.transport.TransportService;
+
+/**
+ * A local implementation of {@link ShardStateAction} that applies shard state changes directly to the
+ * local cluster state. This is used in clusterless mode, where there is no cluster manager.
+ */
+public class LocalShardStateAction extends ShardStateAction {
+    @Inject
+    public LocalShardStateAction(
+        ClusterService clusterService,
+        TransportService transportService,
+        AllocationService allocationService,
+        RerouteService rerouteService,
+        ThreadPool threadPool
+    ) {
+        super(clusterService, transportService, allocationService, rerouteService, threadPool);
+    }
+
+    @Override
+    public void shardStarted(
+        ShardRouting shardRouting,
+        long primaryTerm,
+        String message,
+        ActionListener<Void> listener,
+        ClusterState currentState
+    ) {
+        // We're running in clusterless mode. Apply the state change directly to the local cluster state.
+        ClusterState clusterState = clusterService.state();
+        RoutingTable routingTable = clusterState.getRoutingTable();
+        IndexRoutingTable indexRoutingTable = routingTable.index(shardRouting.index());
+
+        ClusterState.Builder clusterStateBuilder = ClusterState.builder(clusterState);
+        RoutingTable.Builder routingTableBuilder = RoutingTable.builder(routingTable);
+        IndexRoutingTable.Builder indexRoutingTableBuilder = IndexRoutingTable.builder(shardRouting.index());
+        for (IndexShardRoutingTable indexShardRoutingTable : indexRoutingTable) {
+            if (indexShardRoutingTable.shardId().equals(shardRouting.shardId())) {
+                IndexShardRoutingTable.Builder indexShardRoutingTableBuilder = new IndexShardRoutingTable.Builder(indexShardRoutingTable);
+                indexShardRoutingTableBuilder.removeShard(shardRouting);
+                indexShardRoutingTableBuilder.addShard(shardRouting.moveToStarted());
+                indexRoutingTableBuilder.addIndexShard(indexShardRoutingTableBuilder.build());
+            } else {
+                indexRoutingTableBuilder.addIndexShard(indexShardRoutingTable);
```

Comment:
This is required for coordinator level op routing correct?  As in data nodes shouldn't care about shard routing for shards it does not contain.  Can we skip this block if local node is not a coord?